File: src/bin/memaddr.cpp
#include <iostream>
#include <utility>
#include <vector>
using namespace std;

class Addr {
  private:
    static pair<vector<int>, vector<int>> realizer(vector<int> arr,
                                                   vector<int> index) {
        if ((arr.size() % 2 != 0) || (arr.size() / 2 != index.size())) {
            return {};
        }
        vector<int> ra = {};
        vector<int> ri = {};
        for (int i = 0; i < arr.size() / 2; ++i) {
            int curr = i * 2;
            int low = arr[curr];
            int up = arr[curr + 1];
            if (low > up) {
                cout << "Invalid indices" << endl;
                return {};
            }
            ra.push_back(up - low + 1);
            ri.push_back(index[i] - low);
        }
        return {ra, ri};
    }

  public:
    static int rmo(vector<int> arr, vector<int> &index) {
        pair<vector<int>, vector<int>> ai = realizer(arr, index);
        if (ai.first.empty() || ai.second.empty()) {
            return {};
        };
        if (ai.second[0] < 0 || ai.first[0] <= ai.second[0]) {
            return {};
        }
        int offset = ai.second[0];
        for (int i = 1; i < ai.second.size(); ++i) {
            if (ai.second[i] < 0 || ai.first[i] <= ai.second[i]) {
                return {};
            }
            offset = offset * ai.first[i] + ai.second[i];
        }
        return offset;
    }

    static int cmo(vector<int> arr, vector<int> &index) {
        pair<vector<int>, vector<int>> ai = realizer(arr, index);
        if (ai.first.empty() || ai.second.empty()) {
            return {};
        };

        if (ai.second[ai.second.size() - 1] < 0 ||
            ai.first[ai.second.size() - 1] <= ai.second[ai.second.size() - 1]) {
            return {};
        }
        int offset = ai.second[ai.second.size() - 1];
        for (int i = ai.second.size() - 2; i >= 0; --i) {
            if (ai.second[i] < 0 || ai.first[i] <= ai.second[i]) {
                return {};
            }
            offset = offset * ai.first[i] + ai.second[i];
        }
        return offset;
    }
};

int main() {
    vector<int> v = {-5, 5, 7, 20};
    vector<int> i = {3, 10};
    cout << Addr::rmo(v, i) << endl;
    cout << Addr::cmo(v, i) << endl;
}

File: src/bin/memaddr.rs
struct Addr;

impl Addr {
    fn realizer(arr: &[i32], index: &[i32]) -> Option<(Vec<i32>, Vec<i32>)> {
        if arr.is_empty() || !arr.len().is_multiple_of(2) || arr.len() / 2 != index.len() {
            return None;
        }
        let mut ra = Vec::new();
        let mut ri = Vec::new();
        for (i, v) in index.iter().enumerate() {
            let curr: usize = i * 2;
            if arr[curr] > arr[curr + 1] {
                return None;
            }
            if *v < arr[curr] || *v > arr[curr + 1] {
                return None;
            }
            ra.push(arr[curr + 1] - arr[curr] + 1);
            ri.push(v - arr[curr]);
        }
        Some((ra, ri))
    }

    fn rmo(arr: &[i32], index: &[i32]) -> Option<i32> {
        let (ra, ri) = Self::realizer(arr, index)?;
        let mut offset = ri[0];
        for i in 1..ra.len() {
            offset = offset * ra[i] + ri[i];
        }
        Some(offset)
    }

    fn cmo(arr: &[i32], index: &[i32]) -> Option<i32> {
        let (ra, ri) = Self::realizer(arr, index)?;
        let mut offset = ri[ri.len() - 1];
        for i in (0..ra.len() - 1).rev() {
            offset = offset * ra[i] + ri[i];
        }
        Some(offset)
    }
}

fn main() {
    let v = [-5, 5, 7, 20];
    let i = [3, 10];
    let x = Addr::rmo(&v, &i).unwrap();
    println!("{x}");
    let x = Addr::cmo(&v, &i).unwrap();
    println!("{x}");
}
